sdk/lib/rtl/rangelist.c

```diff
@@ -13,16 +13,478 @@
 #define NDEBUG
 #include <debug.h>
 
+/* GLOBALS ******************************************************************/
+
+//extern PAGED_LOOKASIDE_LIST RtlpRangeListEntryLookasideList;
+
 /* TYPES ********************************************************************/
 
-typedef struct _RTL_RANGE_ENTRY
-{
-    LIST_ENTRY Entry;
-    RTL_RANGE Range;
-} RTL_RANGE_ENTRY, *PRTL_RANGE_ENTRY;
+/* RTLP_RANGE_LIST_ENTRY == RTL_RANGE + ListEntry */
+
+/*  FIXME ? ntddk.h */
+/* Flags for public functions: RtlFindRange(), RtlIsRangeAvailable() */
+#define RTL_RANGE_LIST_SHARED_OK           1
+#define RTL_RANGE_LIST_NULL_CONFLICT_OK    2
+/* ? ntddk.h */
+
+/* RTLP_RANGE_LIST_ENTRY.PrivateFlags */
+#define RTLP_ENTRY_IS_MERGED  1
+
 
 /* FUNCTIONS ***************************************************************/
 
+static
+PRTLP_RANGE_LIST_ENTRY
+FASTCALL
+RtlpEntryFromLink(PVOID Link)
+{
+    return CONTAINING_RECORD(Link, RTLP_RANGE_LIST_ENTRY, ListEntry);
+}
+
+static
+BOOLEAN
+FASTCALL
+IsRangesIntersection(
+    _In_ PRTLP_RANGE_LIST_ENTRY Entry1,
+    _In_ PRTLP_RANGE_LIST_ENTRY Entry2)
+{
+    /* Not intersection when:
+       (Entry1->Start - Entry1->End) ... (Entry2->Start -
+       (Entry2->Start - Entry2->End) ... (Entry1->Start -
+    */
+
+    if (((Entry2->Start > Entry1->Start && Entry2->Start > Entry1->End) ||
+         (Entry1->Start > Entry2->Start && Entry1->Start > Entry2->End)))
+    {
+        /* No intersection */
+        return FALSE;
+    }
+    else
+    {
+        /* Entry1 and Entry2 overlap */
+        return TRUE;
+    }
+}
+
+VOID
+NTAPI
+RtlpDeleteRangeListEntry(
+    _In_ PRTLP_RANGE_LIST_ENTRY DelEntry)
+{
+    PRTLP_RANGE_LIST_ENTRY RtlEntry;
+    PRTLP_RANGE_LIST_ENTRY NextRtlEntry;
+
+    PAGED_CODE_RTL();
```

Comment:
The function should have `CODE_SEG("PAGE")` segment then (here and in other functions)

Reply:
Ok